lib/phoenix_live_view/upload_config.ex

```diff
@@ -0,0 +1,570 @@
+    new_metas = Map.put(conf.entry_refs_to_metas, entry.ref, %{})
+
+    %UploadConfig{
+      conf
+      | entries: conf.entries ++ [entry],
+        entry_refs_to_pids: new_pids,
+        entry_refs_to_metas: new_metas
+    }
+  end
+
+  defp put_invalid_entry(conf, entry, reason) do
+    entry = %UploadEntry{entry | valid?: false}
+    new_pids = Map.put(conf.entry_refs_to_pids, entry.ref, @invalid)
+    new_metas = Map.put(conf.entry_refs_to_metas, entry.ref, %{})
+
+    new_conf = %UploadConfig{
+      conf
+      | entries: conf.entries ++ [entry],
+        entry_refs_to_pids: new_pids,
+        entry_refs_to_metas: new_metas
+    }
+
+    put_error(new_conf, entry.ref, reason)
+  end
+
+  defp validate_max_file_size({:ok, %UploadEntry{client_size: size}}, %UploadConfig{
+         max_file_size: max
+       })
+       when size > max or not is_integer(size),
+       do: {:error, :too_large}
+
+  defp validate_max_file_size({:ok, entry}, _conf), do: {:ok, entry}
+
+  defp validate_accepted({:ok, %UploadEntry{} = entry}, conf) do
+    if accepted?(conf, entry) do
+      {:ok, entry}
+    else
+      {:error, :not_accepted}
+    end
+  end
+
+  defp validate_accepted({:error, _} = error, _conf), do: error
+
+  defp accepted?(%UploadConfig{accept: :any}, _entry), do: true
+
+  defp accepted?(%UploadConfig{accept: %{"image/*" => _}}, %UploadEntry{
+         client_type: <<"image/" <> _>>
+       }),
+       do: true
+
+  defp accepted?(%UploadConfig{accept: %{"audio/*" => _}}, %UploadEntry{
+         client_type: <<"audio/" <> _>>
+       }),
+       do: true
+
+  defp accepted?(%UploadConfig{accept: %{"video/*" => _}}, %UploadEntry{
+         client_type: <<"video/" <> _>>
+       }),
+       do: true
+
+  defp accepted?(%UploadConfig{accept: accept}, %UploadEntry{} = entry) do
+    cond do
+      Map.has_key?(accept, entry.client_type) -> true
+      Path.extname(entry.client_name) in (accept |> Map.values() |> Enum.concat()) -> true
+      true -> false
+    end
+  end
+
+  defp recalculate_computed_fields(%UploadConfig{} = conf) do
+    recalculate_errors(conf)
+  end
+
+  defp recalculate_errors(%UploadConfig{ref: ref} = conf) do
+    if too_many_files?(conf) do
+      conf
+    else
+      new_errors =
+        Enum.filter(conf.errors, fn
+          {^ref, @too_many_files} -> false
+          _ -> true
+        end)
+
+      %UploadConfig{conf | errors: new_errors}
+    end
+  end
+
+  @doc false
+  def put_error(%UploadConfig{} = conf, _entry_ref, @too_many_files = reason) do
+    %UploadConfig{conf | errors: Enum.uniq(conf.errors ++ [{conf.ref, reason}])}
+  end
+
+  def put_error(%UploadConfig{} = conf, entry_ref, reason) do
+    %UploadConfig{conf | errors: conf.errors ++ [{entry_ref, reason}]}
+  end
+
+  @doc false
+  def cancel_entry(%UploadConfig{} = conf, %UploadEntry{} = entry) do
+    case entry_pid(conf, entry) do
+      channel_pid when is_pid(channel_pid) ->
+        Phoenix.LiveView.UploadChannel.cancel(channel_pid)
```

Comment:
This is the only place we call the channel in this module, other than this it is completely isolated from channel logic. Shall we move this function to the `Upload` module instead?